client/session.go

```diff
@@ -683,28 +685,19 @@ func (s *session) Write(namespace, id string, t time.Time, value float64, unit x
 
 	for i := range state.queues {
 		if err := state.queues[i].Enqueue(state.op); err != nil {
-			state.Unlock()
-			state.decRef()
-
-			// NB(r): if this ever happens we have a code bug, once we
-			// are in the read lock the current queues we are using should
-			// never be closed
-			s.RUnlock()
-			s.opts.InstrumentOptions().Logger().Errorf("failed to enqueue write: %v", err)
-			return err
+			// NB(r): if this happens we have a bug, once we are in the read
+			// lock the current queues should never be closed
+			panic(fmt.Sprintf("failed to enqueue write: %v", err))
```

Comment:
A better error description is warranted if we want to make this a panic.  Also as an engineer that would prefer to avoid a whole set of ingesters panicing if this case is introduced in a new release, I'd prefer to see the write retried rather than the entire process panicing and taking a while to come back up while the same issue sporadically affects the rest of the fleet.

I think we should leave the non-panic behavior intact.